Fixes connection pool bug

icalderontt opened the pull request:
Now it makes use of a connection pool when necessary which should boost
requests speed.

icalderontt committed:
Fixes connection pool bug

Now it makes use of a connection pool when necessary which should boost
requests speed.

brozeph commented on .nvmrc:
```diff
@@ -0,0 +1 @@
+8.15.0
```
I have philosophical reservations around checking-in developer specific environment artifacts like this `.nvmrc' file. While it is not likely to cause issues, many devs use `tj/n` or even Homebrew (uck...) to manage their Node versions and this file would be vestigial. It's not like checking in a VS Code prefs file... perhaps we should add the .nvmrc to the .gitignore?

icalderontt replied:
I removed it based on your comments. With my previous employer we always committed the .nvmrc file so that the next dev would know what version they are supposed to be running so I'm just used to committing it :D

icalderontt commented:
Here's the related ticket: http://jira.touchtunes.com/browse/BMS-8

brozeph commented on lib/utils.js:
```diff
@@ -0,0 +1,75 @@
+	 */
+	function createAgent(isSecure, options = {}) {
+		return new (isSecure ? https : http).Agent(options);
+  }
```
Generally looks like there there are mixed tabs and spaces in the file (see lines 23 to 26 above)...

icalderontt replied:
whoops. Good find. I fixed it

icalderontt committed:
gitignore now ignores .nvmrc files

cmerchant commented:
Please make sure you guys are updating the the version in package.json file. The last PR for vnext branch did not include an update to the version.

privard commented on lib/utils.js:
```diff
@@ -0,0 +1,75 @@
+ * subsequent calls to the methods so that the same agents are reused for
+ * all requests
+ */
+module.exports.getHTTPAgent = (function () {
```
This file seems to export only one thing, so why not call it getHTTPAgent.js instead of utils.js? Or even, connectionPool.js, or agentManager.js or something more scoped to its actual functionality?

icalderontt replied:
Makes sense man. I just some changes I made based on your comment

icalderontt committed:
refactored http-agent creation code

Added all the relevant code to the http-agent.js file and also updated
the unit tests for this module.

privard reviewed (commented):
The refactoring looks great. Before I approve, I think we need to check into how this is going to scale. I'm wondering how many connections a single agent can make, how how many requests per sec it can make on a single connection. We should definitely find a way to get numbers on this before we move this fix to production.

icalderontt commented:
@privard: that a very valid concern. I couldn't find anything online regarding reusing agents and scaling problems. One thing that we could do if we found any issues is set the max number of allowed sockets in the agent to a # higher than 32. But I think that we should heavily test in dev first

privard commented:
Sounds good, merge at will
